frontend/catalyst/__init__.py

```diff
@@ -91,6 +95,7 @@
     "vjp",
     "jvp",
     "adjoint",
+    "debug_print",
```

Comment:
@dime10 two suggestions here:

- We name the function `catalyst.print` directly, or
- We create a new Python module `catalyst.debug.print`

I quite like the latter (a debug module) since it allows us to have a `debug` section in the API docs with an overview of debugging information, rather than to have a separate `.rst` quickstart doc.

E.g., allows us to keep the docs closed to the code. Similar to how https://docs.pennylane.ai/en/stable/code/qml_shadows.html acts as the main documentation for Shadows in PL.

Reply:
So the reason I didn't name it `catalyst.print` is that it would override the built-in print function if you import it from catalyst.

Reply:
Good point - in that case I am in favour either of `catalyst.debug_print` or `catalyst.debug.print` (assuming here in our examples we import this absolutely, to avoid the overwriting)

Reply:
I'm ok adding it in a `debug` submodule, you just won't be able to (or shouldn't) import it directly to avoid hiding the built-in print function 👍 
I think the reason I avoided it is because in Catalyst we have no (user-facing) submodules (whereas JAX has plenty, e.g. `jax.numpy`, so there it's not unusual).

Reply:
Yep, let's leave it as is for now :+1:

Reply:
@josh146 How do you think this function should be integrated into PL?

Perhaps we can consider one the following UI:
- `qml.compiler.print`
- `qml.compiler.debug_print`

I think the former makes more sense with the design of PL.

Reply:
Yep I'm happy with `qml.compiler.debug_print`

Reply:
Would it not live under `qml.debug_print`, like the other Catalyst functions? iirc the `compiler` module was mostly for developers / utility functions, rather than for user functions?

Reply:
Hmm, maybe. I guess my concern would be - I would love if the name/import path indicated to a user that it is used for printing during runtime within `qjit`. Otherwise, we could see users trying to use it within `@qjit`, because it has such a useful sounding name.

So perhaps something like `qml.runtime_print`, or `qml.jit_print`?

Reply:
imo the name `debug_print` implies a compiler related utility function, and as such, it may not be an issue if users want to use it from `qml.compiler`.

`qml.runtime_print` or `qml.jit_print` makes sense if we want to extend the support for `jax.jit` in PL.

Reply:
Yep @maliasadi I agree. `qml.compiler.debug_print` works for me because of the namespacing

Reply:
> But is that a problem? All our functions (should) work even outside of qjit, so a user could certainly use it in a Python context.

oh yeah 🤦🤦🤦🤦🤦

Reply:
>  Rather, it is a program instruction, just like catalyst.grad or qml.Hadamard.

I'm sorry, but this is not completely right! Based on the objectives of this work, the `debug_print` is a "printf-style debugging to catalyst." And as such we should treat it as a debugging functionality and it defers from other program instructions that we have.

> Something like qml.runtime_print seems reasonable, we should just ensure that it is consistent in the Catalyst package as well.

What about renaming `debug_print` to `runtime_print` in Catalyst?

Reply:
> I'm sorry, but this is not completely right! Based on the objectives of this work, the debug_print is a "printf-style debugging to catalyst." And as such we should treat it as a debugging functionality and it defers from other program instructions that we have.

What you're saying doesn't contradict my statement, it is not a compiler function, but a program instruction.

Reply:
of course it doesn't! I was emphasizing the difference between this debugger function and other Catalyst program instructions we have implemented so far! I insist this should be treated differently and I'm suggesting to integrate it in the PLCF work as a compiler/debugging method.

I don't have a strong opinion against renaming it as `qml.runtime_print` as part of `qjit_api` if you both are happy with it. But, why not renaming this method here in Catalyst to `catalyst.runtime_print`?

Reply:
After further deliberation we decided on `catalyst.debug.print`, so I'm closing this thread.